src/main/java/com/google/devtools/build/lib/cmdline/PackageIdentifier.java

```diff
@@ -177,6 +178,65 @@ public String getCanonicalForm() {
     return repository.getCanonicalForm() + "//" + getPackageFragment();
   }
 
+  /**
+   * Returns an absolutely unambiguous canonical form for this package in label form. Parsing this
+   * string in any environment, even when subject to repository mapping, should identify the same
+   * package.
+   */
+  public String getUnambiguousCanonicalForm() {
+    return String.format("@@%s//%s", getRepository().getName(), getPackageFragment());
+  }
+
+  /**
+   * Returns a label representation for this package that is suitable for display. The returned
+   * string is as simple as possible while referencing the current package when parsed in the
+   * context of the main repository.
+   *
+   * @param mainOrOtherRepoMapping the {@link RepositoryMapping} of the main repository or, if that
+   *                               is not available to the caller, the {@link RepositoryMapping} of
+   *                               the current package. The former will yield more readable output.
+   * @return <ul><li>if the main repository's mapping is provided:</li>
+   * <dl>
+   * <dt><code>//some/pkg</code></dt>
+   * <dd>if this package lives in the main repository</dd>
+   * <dt><code>@protobuf//some/pkg</code></dt>
+   * <dd>if this package lives in a repository with "protobuf" as <code>name</code> of a repository
+   * in WORKSPACE or as local name of a Bzlmod dependency of the main module</dd>
+   * <dt><code>@@protobuf~3.19.2//some/pkg</code></dt>
+   * <dd>only with Bzlmod if the current package belongs to a repository that is not visible from
+   * the main module</dd>
+   *
+   * <li>if another repository's mapping is provided:</li>
+   * <dl>
+   * <dt><pre>//some/pkg</pre></dt>
+   * <dd>if this package lives in the main repository</dd>
+   * <dt><pre>@protobuf//some/pkg</pre></dt>
+   * <dd>if this package lives in a repository with "protobuf" as <code>name</code> of a repository
+   * in WORKSPACE or as local name of a Bzlmod dependency of the main module</dd>
```

Comment:
"or as local name of a Bzlmod dependency of the main module" does not apply here, does it?

Reply:
Fixed.